remove "default" CaseFormat for stringConstants!

dglachs opened the pull request:
this solves the following problem:

when having similar names for attributes and classes, such as
- ssn:FeatureOfInterest (denoting a class)
- ssn:featureOfInterest (denoting a property)

having UPPER_UNDERSCORE as the default stringConstantCase will cause duplicate fields ...and there is no possibility to omit reformatting to any of the CaseFormat options.

dglachs committed:
remove "default" CaseFormat for stringConstants!

this solves the following problem:

when having similar names for attributes and classes, such as

- ssn:FeatureOfInterest (denoting a class)
- ssn:featureOfInterest (denoting a property)

reformatting the case will cause duplicate fields ... but with the
default value it was not possible to omit the case format field!

ansell commented:
Are you able to modify the maven property to do what you need without changing the default? 

The reason that the default case format exists is to make it work for the majority of cases that we need it to work for, while making sure that others are able to reconfigure it themselves in their pom.xml files.

dglachs commented:
Hello Peter, Thomas,

the problem I see, that there is no valid option in com.google.common.base.CaseFormat which says "leave as it is". In fact, for now, there are the following options:

1. omitting the tag <stringConstantCase>: this will do the default of UPPER_UNDERSCORE 
2. provide the tag but with no value: this will do the default of UPPER_UNDERSCORE
3. use another valid CaseFormat option (LOWER_HYPHEN, UPPER_HYPHEN etc.): this will do the requested case adjustments but result in a duplicate field (at least for the SSN ontology as mentioned in my request and also for the FOAF ontology where for example foaf:familyName and foaf:family_name cause a conflict)
4. use an invalid option such as DO_NOTHING: this will result in an error since the configuration cannot be parsed.

For me, the simplest solution was to remove the default, because this results in a "no case format adjustment" when omitting the <stringConstantCase> tag. 

However, a more sophisticated solution could allow
- specify the default generation options for URI's and/or strings including case format, prefix, suffix 
- allow overriding the default settings for each configured vocabulary (partially already done with vocabulary-specific caseFormat) 

Best
  Dietmar

ansell commented:
Thanks, that explains it quite well I think. In that case, I would definitely support offering either a dummy value or default to nothing and the current users update our pom.xml files with our desired cases.

dglachs committed:
Helper class to specify 
- prefix
- suffix
- caseFormat adjustment

dglachs committed:
Added elements 

- uriGeneration (vocabulary specific prefix& suffix)
- stringGeneration (vocabulary specific prefix & suffix)

dglachs committed:
adopted to have 

- default generation settings for URI and String
- allow vocabulary specific settings which override default generation

dglachs committed:
Adopted to use generation settings for URI & String

dglachs committed:
Adjusted to use the GenerationSetting for URI & String generation

no changes in parameter list

dglachs committed:
stringConstantCase is now unused!!!

the parameter is still here in order to avoid parse exceptions
(depending on pom.xml configuration)

dglachs commented:
there was some time left so I modified the functionality quite a bit in order to provide more (generation) flexibility: 

(1) allow to specify prefix & suffix for the URI constants as well. 
either provide the tags in the plugin as 

    <createUriConstants>true</createUriConstants>    (defaults to true)
    <uriConstantPrefix></uriConstantPrefix> (defaults to "")
    <uriConstantSuffix></uriConstantSuffix> (defaults to "")
    <createStringConstants>true</createStringConstants>    (defaults to true)
    <stringConstantPrefix></stringConstantPrefix> (defaults to "")
    <stringConstantSuffix></stringConstantSuffix> (defaults to "_STRING")

or in an special "setting" structure (which may also be used for each vocabulary)

    <uriGeneration>   (or <stringGeneration>)
    ...<caseFormat></caseFormat>
    ...<constantPrefix></constantPrefix>
    ...<constantSuffix></constantSuffix> 
    </uriGeneration>

this for example gives you the possibility to configure a "_URI" suffix for URI's (if desired) . in addition, vocabulary specific settings are also possible. 

(2) default case format will be used for both URI & String generation...
I personally found it not logical to have URI constants adjusted differentlty than String constants by default. So I'd propose to have only one "caseFormat" element which is used for both generation options by default. now, no caseFormat adjustment for URI's takes place by default (as it was before). the caseFormat setting is then used for string generation as well, unless it is further specified in a &lt;stringGeneration&gt; element.

(3) no changes in pom.xml of your users required, command line parameters are still the same
the changes are just additions, nothing has been removed.

finally, there were quite some adjustments and changes in the maven plugin and in the VocabBuilder class as well. Maybe you find the changes useful.

dglachs committed:
corrected check

ja-fra commented:
I'd suggest to switch to the new configuration format @dglachs suggests and deprecate the old way (to be removed in one the next versions). Keeping both is too much overhead and will confuse users.

For the defaults: I'd add a special handling for `AS_IS` or `NONE` which then should be also used as default/not explicitly set...

And finally: After merging these changes we should have new release!

tkurz commented:
I totally agree with @ja-fra . @dglachs could you add the 'NONE' case? Afterwards we can merge and release.

dglachs committed:
Provide caseFormat-Option "NONE"

Changes to support the NONE option for caseFormat adjustments.

dglachs commented:
Added the required functionality ... 

currently the following tags in the &lt;configuration&gt; section are possible:

```
<caseFormat>NONE</caseFormat>
```

this is the default and specifies NO caseFormat-Adjustments for both URI & String-Constant generation. (the distinct default generation options are still specified with the "old" parameters, see my previous comment). For having more fine grained generation options - the GenerationSetting class is used. One may specify

```
<uriGeneration>   (or <stringGeneration>)
    <caseFormat></caseFormat>
    <constantPrefix></constantPrefix>
    <constantSuffix></constantSuffix> 
</uriGeneration>
```

which then override the default-settings. to have it more sophisticated it is also possible to have the following structure along with a vocabulary (under &lt;vocabularies&gt;)

```
<vocabulary>
    <className>FOAF</className>
    <prefix>http://xmlns.com/foaf/0.1/</prefix>
    <file>${project.build.directory}/../ontologies/foaf.ttl</file>
    <uriGeneration>   (or <stringGeneration>)
        <caseFormat></caseFormat>
        <constantPrefix></constantPrefix>
        <constantSuffix></constantSuffix> 
    </uriGeneration>
</vocabulary>
```

for specifying specific prefix/suffix for a particular vocabulary (overriding the general settings from above). 

As noted - there are still the "duplicated" or "old" parameters (&lt;stringConstantPrefix&gt;,&lt;stringConstantSuffix&gt; etc.) - I've marked them as deprecated. Anyways, as @ja-fra suggested, these "old" parameters should be removed, for this, it is recommended to remove only the Parameter-Annotation and just leave the variables including their default as constants.

ansell commented on vocab-builder-cli/src/main/java/com/github/tkurz/sesame/vocab/Main.java:
```diff
@@ -100,38 +100,82 @@ public static void main(String[] args) {
+            final GenerationSetting uriGenerationSetting = new GenerationSetting(); 
             if (cli.hasOption('c')) {
+            	String uriCaseFormat = cli.getOptionValue('c');
+            	uriGenerationSetting.setCaseFormat(uriCaseFormat);
```
This is an odd style of programming to optimistically set the case format and then check it later and throw an exception. Typically, to make code maintainable, the setter is only called after it is validated (or the validation happens inside of the setter).

ansell commented on vocab-builder-cli/src/main/java/com/github/tkurz/sesame/vocab/Main.java:
```diff
@@ -100,38 +100,82 @@ public static void main(String[] args) {
-            } else {
-                builder.setStringPropertyPrefix(null);
-            }
+            CaseFormat caseFormat = null;
```
This does not need to be a shared variable between "c" and "C" parsing. It doesn't even look like it needs to be stored, as it is only used for a null check in both locations and the string is what is passed into GenerationSetting now instead of the CaseFormat object.

Is it redundant to pass the string around instead of passing a CaseFormat object around?

ansell commented on vocab-builder-cli/src/main/java/com/github/tkurz/sesame/vocab/Main.java:
```diff
@@ -100,38 +100,82 @@ public static void main(String[] args) {
+                if (cli.hasOption('C')) {
+                	setting.setCaseFormat(cli.getOptionValue('C'));
+                    try {
+                        final CaseFormat strCaseFormat = CaseFormat.valueOf(cli.getOptionValue('C'));
```
Ditto comment above.

ansell commented on vocab-builder-cli/src/main/java/com/github/tkurz/sesame/vocab/Main.java:
```diff
@@ -100,38 +100,82 @@ public static void main(String[] args) {
+                	setting.setCaseFormat(cli.getOptionValue('C'));
+                    try {
+                        final CaseFormat strCaseFormat = CaseFormat.valueOf(cli.getOptionValue('C'));
+                        if (caseFormat == null) {
```
This looks like it is checking the wrong variable, as a result of pushing the caseFormat variable into the main method scope.

ansell commented on vocab-builder-cli/src/main/java/com/github/tkurz/sesame/vocab/Main.java:
```diff
@@ -100,38 +100,82 @@ public static void main(String[] args) {
+                // specify string generation
+            	builder.setStringGeneration(setting);
             }
+//            if (cli.hasOption('S')) {
```
Just delete the following code rather than commenting it. It is always available if we need it in the git repository.

ansell commented on vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/GenerationSetting.java:
```diff
@@ -0,0 +1,53 @@
+	private String constantSuffix;
+	public CaseFormat getCaseFormat() {
+		try {
+			if ( caseFormat != null ) {
```
No need for this null check if we are catching all exceptions. CaseFormat.valueOf will throw an NPE for that.

Alternatively, the null check could be outside of the try-catch block to emphasise that the exceptions are expected from CaseFormat.valueOf

ansell commented on vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/GenerationSetting.java:
```diff
@@ -0,0 +1,53 @@
+		return null;
+	}
+	public void setCaseFormat(String caseFormat) {
+		this.caseFormat = caseFormat;
```
It would be useful to do validation here rather than waiting until the get method is called to validate the input.

dglachs replied:
yes - will change it

ansell commented on vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/GenerationSetting.java:
```diff
@@ -0,0 +1,53 @@
+		
+	}
+	
+	public GenerationSetting mergeWith(GenerationSetting other) {
```
Need javadoc to specify the contract here, as it isn't clear what semantics the merge operation has without reading the code.

ansell commented:
Overall, the code also needs to be reformatted to use spaces instead of tabs.

ansell commented on vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/VocabBuilder.java:
```diff
@@ -190,73 +197,150 @@ public void generate(String className, PrintWriter out) throws IOException, Grap
+	        out.println();
+
+		}
+        //string constant values
```
Delete the following commented out code, and after reformatting the code above to use spaces, git will show what the changes you made to this section are.

ansell commented on vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/VocabBuilder.java:
```diff
@@ -296,6 +380,10 @@ public void generateResourceBundle(String baseName, Path bundleDir) throws Gener
 
     public HashMap<String, Properties> generateResourceBundle(String baseName) throws GenerationException {
+    	// be sure to have at least the default URI constant settings
+    	if ( uriGeneration == null ) {
```
The code style we have used here so far is not to have spaces inside of the if statement brackets. Ie, change to:

```
if (uriGeneration == null) {
```

dglachs replied:
ups

ansell commented on vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/VocabBuilder.java:
```diff
@@ -317,7 +405,8 @@ public void generateResourceBundle(String baseName, Path bundleDir) throws Gener
         for (String key : keys) {
             final URI resource = splitUris.get(key);
-            String nextKey = cleanKey(doCaseFormatting(key, getConstantCase()));
+            // 
```
Was this meant to have a comment on this line?

ansell commented on vocab-builder-core/src/main/java/com/github/tkurz/sesame/vocab/VocabBuilder.java:
```diff
@@ -485,34 +574,51 @@ public String getPreferredLanguage() {
     public CaseFormat getConstantCase() {
         return caseFormat;
     }
+//
```
As above, delete code instead of commenting it out.

ansell commented on vocab-builder-maven-plugin/src/main/java/com/github/tkurz/sesame/vocab/plugin/VocabularyBuilderMojo.java:
```diff
@@ -245,19 +313,50 @@ public void execute() throws MojoExecutionException, MojoFailureException {
                         target = target.resolve(builder.getPackageName().replaceAll("\\.", "/"));
                         Files.createDirectories(target);
                     }
+                    if ( createUriConstants && uriGeneration != null ) {
```
This check negates the safety above to "be sure to generate at least the URI's".

ansell replied:
Sorry, should have read down further first.

ansell replied:
Nope, it does actually negate the global safety. You need to reorganise this code and document why the merge would not happen with vocab, but it would happen in this case. Naively, although it is hard to follow your code, I would have thought you would always want to merge the global settings into the vocab specific settings.

dglachs replied:
No, the global settings will be overridden with the vocabulary specific configuration - if set.

ansell commented on vocab-builder-maven-plugin/src/main/java/com/github/tkurz/sesame/vocab/plugin/VocabularyBuilderMojo.java:
```diff
@@ -245,19 +313,50 @@ public void execute() throws MojoExecutionException, MojoFailureException {
+                    	 */
+                    	builder.setStringGeneration(vocab.getStringGeneration());
+                    }
+//                    if ( uriGeneration != null ) {
```
Delete code instead of commenting it out so git can show the full changes

ansell commented on vocab-builder-maven-plugin/src/main/java/com/github/tkurz/sesame/vocab/plugin/VocabularyBuilderMojo.java:
```diff
@@ -314,6 +413,19 @@ public void execute() throws MojoExecutionException, MojoFailureException {
+     * @param caseFormat
+     * @return
+     */
+    private CaseFormat caseFormatFromString(String caseFormat) {
```
This code should not be replicated both here and in GenerationSetting. Need to have a static method in a helper class so that we ensure that we are doing the same operations everywhere for this case.

dglachs committed:
Revert ab4af42..6fb0df3

This rolls back to commit ab4af428699c9d4f148b15170d80a42f5382e088.

dglachs committed:
code cleanup

removed commented sections
added javadoc

dglachs commented:
reverted the NONE option commit since it created more problems than it solved ...

in fact the NONE option was present before because simply omitting the caseFormat-tag (at any level) was the NONE option.  

anyways the proposed configuration structure is still valid

tkurz commented:
The code seems to be okay now. Should we merge the request and release a new version?

ansell commented:
Yes, the functionality looks good, just some style issues remaining that we can automatically reformat after merging and delete some of the commented code that is still present.

Can we switch on TravisCI to do automated tests btw? I thought it was on because we have the .travis.yml file, but it doesn't seem to be enabled at the other end right now:

https://travis-ci.org/tkurz/sesame-vocab-builder

dglachs committed:
code cleanup
